Create gPRC HttpClient project and HttpClientFactory integration

JamesNK opened the pull request:
Addresses https://github.com/grpc/grpc-dotnet/issues/134

* Adds John's gRPC HttpClient implementation as `Grpc.NetCore.HttpClient` project
* Adds HttpClientFactory integration as `Grpc.AspNetCore.Server.GrpcClient` project. @rynowak it would be great if you lend your expert eyes to this area

This PR contains experimental and prototype code but does not touch Grpc.AspNetCore.Server.

There are many many issues to decide on and clean up in future PRs:

* Tests. There is one integration test. Need more
* Split client types out from Grpc.Core to Grpc.Core.Api and change `Grpc.NetCore.HttpClient`'s package reference
* `HttpClientCallInvoker` has some issues: exceptions are not thrown back to the caller, I noticed duplex content was not returned until the caller was closed
* Update HttpClientCallInvoker to use trailer headers. Blocked on HttpClient updates
* c-core gRPC clients are thin wrappers over a channel, and the channel is disposed of. Need to figure out what the life-cycle of the HttpClient will be when used with a gRPC client.
* Figure out what server values can be automatically propagated to a client in the HttpClientFactory. Deadline, cancellation token, propagation token. Are there others?

JamesNK commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcClientServiceExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+            // Accessing the client options here allows the HttpContextAccessor to be added as needed
+            // Is there a way to get the value from services.Configure now?
+            var clientOptions = new GrpcClientOptions<TClient>();
+            configureClient(clientOptions);
```
Is there a better way of doing this? I want to register the configuration with DI - `services.Configure(configureClient);` - and I want to access the configuration immediately.

rynowak replied:
No there really isn't a way to do this, I'd recommend against running the user code here if possible. What's the core concern? 

I have a theory that this is about capturing values from the current request to use in header propagation? Using the context accessor for this is problematic because nothing enforces the constraint that you're "in a request" or "on the request processing thread.

JamesNK replied:
I wanted to access the config value here so the value could be used to decided whether AddHttpContextAccessor was added. If you want to propagate something from the current request then `AddHttpContextAccessor` would be called, otherwise no. Make it pay for play.

rynowak replied:
Considering that the things you check for are the defaults, do you really need this?

JamesNK commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcClientServiceExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+            // HttpContextAccessor has performance overhead. Only add it when required
+            if (RequireHttpContextAccessor(clientOptions))
+            {
+                services.AddHttpContextAccessor();
```
I have heard that `AddHttpContextAccessor` introduces overhead to each request so I made it conditionally added depending on whether the client will use values from the current request. Is this a true fact?

rynowak replied:
It is a true fact. I want to know more about what this is for.

davidfowl replied:
Yes there are also thread safety concerns. This is about cancellation propagation.

JamesNK replied:
Microservices tend to call other microservices. A gRPC client using the current requests cancellation token will allow its calls to be cancelled when the current request is cancelled.

Cancellation token, deadline and context propagation token are values that could be set from the current request.

JamesNK commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcHttpClientFactory.cs:
```diff
@@ -0,0 +1,98 @@
+            }
+
+            // TODO(JamesNK): Need to set deadline and context propagation token
+            // Either add HttpContextServerCallContext to HttpContext.Items or provide equivilent of IHttpContextAccessor
```
Thoughts here? After adding this comment it occurred to me that adding it as a feature to HttpContext might be a better solution.

rynowak replied:
feature is the most first-class thing. Also see: https://github.com/aspnet/AspNetCore/pull/7921

rynowak replied:
If you do end up doing something like this, you need to capture the values up front to avoid safety problems.

JamesNK replied:
Ok! Will figure it out when the todo is looked at.

JamesNK commented on src/Grpc.NetCore.HttpClient/GrpcClientFactory.cs:
```diff
@@ -0,0 +1,72 @@
+        public static TClient Create<TClient>(string baseAddress, X509Certificate certificate = null) where TClient : ClientBase<TClient>
+        {
+            // Needs to be set before creating the HttpClientHandler
+            AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2Support", true);
```
Will we always need this? Will it have an affect on other HttpClient instances in the application?

rynowak replied:
Yes, these are global. I imagine you won't always need it.

JunTaoLuo replied:
This will be needed until the BCL turn this on by default. I don't think that's happening until at least the trailer support is available. cc @shirhatti.

JamesNK commented on test/FunctionalTests/Infrastructure/GrpcTestFixture.cs:
```diff
@@ -47,6 +47,8 @@ public GrpcTestFixture()
 
                 services.AddSingleton<ILoggerFactory>(LoggerFactory);
+
+                services.AddSingleton<IPrimaryMessageHandlerProvider, TestPrimaryMessageHandlerProvider>(s => new TestPrimaryMessageHandlerProvider(_server));
```
Injecting the TestServer into the TestServer

![image](https://user-images.githubusercontent.com/303201/53913849-28ddc880-40c1-11e9-83e7-0988d9f5122c.png)

jtattermusch commented on examples/Clients/Chatter/Program.cs:
```diff
@@ -39,10 +43,8 @@ static async Task<int> Main(string[] args)
-            var channel = new Channel("localhost:50051", credentials);
-            var client = new Chatter.ChatterClient(channel);
+            var certificate = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? null : new X509Certificate2(Path.Combine(Resources.CertDir, "client.crt"));
+            var client = GrpcClientFactory.Create<Chatter.ChatterClient>(@"https://localhost:50051", certificate);
```
Does it make sense to replace existing examples (that are fully working) with examples that use Grpc.NetCore.HttpClient and miss important functionality (trailers)?
It sounds more like we should fork one of the clients and create a separate example for Grpc.NetCore.HttpClient.

Same for other examples.

JamesNK replied:
I've reverted the clients to use the c-core client.

jtattermusch commented on examples/Clients/Chatter/Program.cs:
```diff
@@ -39,10 +43,8 @@ static async Task<int> Main(string[] args)
-            var channel = new Channel("localhost:50051", credentials);
-            var client = new Chatter.ChatterClient(channel);
+            var certificate = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? null : new X509Certificate2(Path.Combine(Resources.CertDir, "client.crt"));
+            var client = GrpcClientFactory.Create<Chatter.ChatterClient>(@"https://localhost:50051", certificate);
```
In most gRPC languages, there's a concept of Channel (something heavyweight, roughly corresponding to a connection even though this parallel is not accurate due to channel laziness and e.g. loadbalancing) and a lightweight "stub" (the client class for given service).
We should consider exposing such concepts too if possible (Channel would be replaced by something that implements a CallInvoker).

JamesNK replied:
Figuring out this probably extends beyond the scope of this PR.

rynowak commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcHttpClientFactory.cs:
```diff
@@ -0,0 +1,98 @@
+        // The Cache should be registered as a singleton, so it that it can
+        // act as a cache for the Activator. This allows the outer class to be registered
+        // as a transient, so that it doesn't close over the application root service provider.
+        public class Cache
```
workarounds, amirite?

JamesNK replied:
I just copied and pasted the default implementation and modified as needed. I assume you know what you're doing here 😄

rynowak commented on src/Grpc.AspNetCore.Server.GrpcClient/Grpc.AspNetCore.Server.GrpcClient.csproj:
```diff
@@ -0,0 +1,15 @@
+    <ProjectReference Include="..\Grpc.NetCore.HttpClient\Grpc.NetCore.HttpClient.csproj" />
+    <PackageReference Include="Grpc.Core" Version="$(GrpcCorePackageVersion)" />
+
+    <PackageReference Include="Microsoft.Extensions.Http" Version="3.0.0-preview.19074.2" />
```
Is this just temporary? This should be part of the framework reference.

rynowak commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcClientOptions.cs:
```diff
@@ -0,0 +1,32 @@
+        public Uri BaseAddress { get; set; }
+        public X509Certificate Certificate { get; set; }
+        public bool UseRequestCancellationToken { get; set; } = true;
+        public bool UseRequestDeadline { get; set; } = true;
```
nit: I find it much less surprising for bool options to default to `false`. It's less confusing (IMO) from the point of view of the app developer.

```
options.SupressRequestCancellationToken = true;
```

Ignore me if this project is already using a different style.

JamesNK replied:
Sure, something to think about. Need to figure out what the default behavour should be before locking down on the names.

rynowak commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcClientServiceExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+{
+    public static class GrpcClientServiceExtensions
+    {
+        public static IHttpClientBuilder AddGrpcClient<TClient>(this IServiceCollection services, Action<GrpcClientOptions<TClient>> configureClient) where TClient : ClientBase<TClient>
```
You might want another overload that takes a `string name`. Typed clients use `nameof(TClient)` as the identifier for the client, so if you need to register multiple types with the same class name, or register the same class name multiple times you will be in trouble.

JamesNK replied:
How do named typed clients work with constructor DI injection?

I'm going to guess that the named configs are usable if IHttpClientFactory is injected into a class and you create an instance from the factory.

JunTaoLuo commented on src/Grpc.NetCore.HttpClient/Grpc.NetCore.HttpClient.csproj:
```diff
@@ -0,0 +1,27 @@
+  <PropertyGroup>
+    <Authors>The gRPC Authors</Authors>
+    <Copyright>Copyright 2019 The gRPC Authors</Copyright>
+    <Description>gRPC support for ASP.NET Core</Description>
```
Managed .NET client for gRPC

JunTaoLuo commented on src/Grpc.NetCore.HttpClient/HttpClientCallInvoker.cs:
```diff
@@ -0,0 +1,400 @@
+        }
+    }
+
+    internal class PipeClientStreamWriter<TRequest> : IClientStreamWriter<TRequest>
```
These files need to be split up.

JamesNK commented on src/Grpc.AspNetCore.Server.GrpcClient/Grpc.AspNetCore.Server.GrpcClient.csproj:
```diff
@@ -0,0 +1,15 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
It has a dependency on the managed gRPC client. We don't nessessarily want the server to always bring in the client.

This package glues the two together. I think it would be something the metadata would bring in.

JamesNK commented:
🆙 📅 

I have reverted the example clients to use the c-core client. All changes are in two new projects. I'd like to get this in, and we can production-ize the managed gRPC client in later PRs.

JamesNK commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcHttpClientFactory.cs:
```diff
@@ -0,0 +1,98 @@
+            }
+
+            var callInvoker = new HttpClientCallInvoker(httpClient);
+            if (_clientOptions.UseRequestCancellationToken)
```
@rynowak I need to access named `GrpcClientOptions` options here but I don't have the name. Is this a limitation of `ITypedHttpClientFactory<TClient>`?

JamesNK commented:
🆙 📅 

* Added named client configuration
* Added `IGrpcClientFactory` for resolving clients by name
* Added `IServerCallContextFeature` for accessing context from `HttpContext`
* Request deadline/cancellation token passed to client
* Unit tests

This (the HttpClientFactory integration) is feature complete and has tests. I'd like to get this in so please review 🙏

jtattermusch commented on src/Grpc.NetCore.HttpClient/Grpc.NetCore.HttpClient.csproj:
```diff
@@ -0,0 +1,26 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Should we put a README.md under Grpc.NetCore.Http client that this is not expected to work yet?

Status and Trailers are not supported yet, so calls will always return StatusCode.OK, so (unlike Grpc.AspNetCore.Server) the managed client is not really usable yet. We should potential early adopters about this.

JamesNK replied:
Good idea

jtattermusch commented:
travis tests seem stuck, not sure why. Rebase to nudge travis a bit?

JamesNK commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcClientOptions.cs:
```diff
@@ -0,0 +1,53 @@
+        /// <summary>
+        /// The base address to use when making gRPC requests.
+        /// </summary>
+        public Uri BaseAddress { get; set; }
```
Yes you could. I like `BaseAddress` because it matches what developers are fimilar with from `HttpClient.BaseAddress`

JamesNK commented on src/Grpc.AspNetCore.Server.GrpcClient/TypeNameHelper.cs:
```diff
@@ -0,0 +1,195 @@
+
+namespace Grpc.AspNetCore.Server.GrpcClient
+{
+    // Copied from https://github.com/aspnet/Extensions/blob/aa18b93adac411b6b98e4088284a1ffb7d83249f/src/Shared/src/TypeNameHelper/TypeNameHelper.cs
```
The history of this code is it is entirely written by Microsoft developers. We're contributing it to Grpc like any other code we are writing for the project.

If you still have concerns then we can chat.

jtattermusch replied:
Got that, my concern something along the line that automated license checkers won't know that and might evaluate as false positive.

JamesNK commented on src/Grpc.NetCore.HttpClient/ClientAsyncStreamReader.cs:
```diff
@@ -0,0 +1,66 @@
+            }
+            catch (Exception ex)
+            {
+                ex.StackTrace.ToString();
```
Whoops, old code from trying to debug the client. Will remove.

JunTaoLuo commented on build/dependencies.props:
```diff
@@ -9,9 +9,10 @@
     <NunitPackageVersion>3.11.0</NunitPackageVersion>
     <Nunit3TestAdapterPackageVersion>3.12.0</Nunit3TestAdapterPackageVersion>
-    <MicrosoftSourceLinkGitHubPackageVersion>1.0.0-beta2-18618-05</MicrosoftSourceLinkGitHubPackageVersion>
+    <SystemIOPipelinesPackageVersion>4.6.0-preview.19073.11</SystemIOPipelinesPackageVersion>
```
4.6.0-preview3.19128.7 looks like the newest version.

JamesNK replied:
Pipelines will eventually be removed from the client so not important.

JunTaoLuo commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcClientServiceExtensions.cs:
```diff
@@ -0,0 +1,192 @@
+            return builder;
+        }
+
+        internal class DefaultHttpClientBuilder : IHttpClientBuilder
```
nit: private class, or new files

JunTaoLuo commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcHttpClientFactory.cs:
```diff
@@ -0,0 +1,115 @@
+
+                callInvoker.CancellationToken = serverCallContext.CancellationToken;
+            }
+            if (namedOptions.PropagateDeadline)
```
nit: new line after brace

JunTaoLuo commented on src/Grpc.AspNetCore.Server.GrpcClient/GrpcHttpClientFactory.cs:
```diff
@@ -0,0 +1,115 @@
+        // The Cache should be registered as a singleton, so it that it can
+        // act as a cache for the Activator. This allows the outer class to be registered
+        // as a transient, so that it doesn't close over the application root service provider.
+        public class Cache
```
Why is this public?

JamesNK replied:
It's nested in an internal class so it becomes internal anyway.

JunTaoLuo commented on src/Grpc.AspNetCore.Server.GrpcClient/INamedTypedHttpClientFactory.cs:
```diff
@@ -0,0 +1,27 @@
+
+using System.Net.Http;
+
+namespace Grpc.AspNetCore.Server.GrpcClient
```
Internal namespace? Also, move to internal directory?

rynowak commented on src/Grpc.AspNetCore.Server.GrpcClient/DefaultGrpcClientFactory.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace Grpc.AspNetCore.Server.GrpcClient
+{
+    internal class DefaultGrpcClientFactory : IGrpcClientFactory
```
I was going to complain about this because of parallel abstractions - but I think it seems fine since a Grpc client already is a different type from httpclient.

rynowak commented on src/Grpc.AspNetCore.Server.GrpcClient/IGrpcClientFactory.cs:
```diff
@@ -0,0 +1,37 @@
+    /// A factory abstraction for a component that can create gRPC client instances with custom
+    /// configuration for a given logical name.
+    /// </summary>
+    public interface IGrpcClientFactory
```
Consider seriously making this an abstract class. We should have done with `IHttpClientFactory` .

rynowak commented on src/Grpc.AspNetCore.Server/Internal/CallHandlers/ServerCallHandlerBase.cs:
```diff
@@ -18,6 +18,7 @@
 
 using System;
 using System.Threading.Tasks;
+using Grpc.AspNetCore.Server.Feature;
```
Is there a really good reason to create a namespace for this?

JamesNK replied:
Nothing other than features everywhere else live in a Features namespace.

I'm going to leave it here for now and we can change it in an API review closer to release.

rynowak replied:
IMO this naming doesn't make sense in our code, but is excusable because we have lots of features. I can't imagine grpc having too many, but maybe I need to adopt a growth mindset!
